### PR TITLE
topic/T318-086

### PR DESCRIPTION
**Pull request by joaopsazevedo**
Suppress Separate refactoring tool.
Merging into edge since it needs laltools master.

### CODE REVIEW

File: source/ada/lsp-lal_utils.adb [hunk 1]
```diff
@@ -492,6 +492,116 @@ package body LSP.Lal_Utils is
...
+   is
+      File_URI : LSP.Types.LSP_String;
+
+      Text_Edits : LSP.Messages.TextEdit_Vector;
+
+      use Laltools.Refactor;
+
+      Text_Edits_Cursor     : Text_Edit_Ordered_Maps.Cursor
+        := Edits.Text_Edits.First;
+      --  File_Creations_Cursor : File_Creation_Ordered_Sets.Cursor
```
Comment (reznikmm) [inline source/ada/lsp-lal_utils.adb:515]: Can you drop commented code?

File: source/protocol/lsp-messages.ads [hunk 2]
```diff
@@ -1254,35 +1254,65 @@ package LSP.Messages is
       ignoreIfExists : Optional_Boolean;
    end record;
 
+   procedure Write_CreateFileOptions
```
Comment (reznikmm) [inline source/protocol/lsp-messages.ads:1257]: It would be nice to have Read subprograms also.

File: source/protocol/lsp-messages.adb [hunk 3]
```diff
@@ -1121,6 +1121,58 @@ package body LSP.Messages is
...
+      V : LSP.Messages.CreateFile)
+   is
+      JS : LSP.JSON_Streams.JSON_Stream'Class renames
+        LSP.JSON_Streams.JSON_Stream'Class (S.all);
+      Kind : constant LSP.Types.LSP_String := LSP.Types.To_LSP_String
+        (Ada.Strings.UTF_Encoding.UTF_8_String'("create"));
+   begin
+      JS.Start_Object;
+      JS.Key ("kind");
+      LSP.Types.Write (S, Kind);
```
Comment (reznikmm) [inline source/protocol/lsp-messages.adb:1135]: Perhaps we can add 'kind' fields to CreateFile/DeleteFile/RenameFile record types and use auto-generated Read/Write subprograms.
```ada
kind : Document_Change_Kind range Create_File .. Create_File := Create_File;
```

Not sure if we shall use `Document_Change_Kind` or another dedicated enumeration type. I can do this if you want.

**Review by reznikmm — APPROVED** [review]
* Why not merge this into `master`?
* Can we have any documentation of available refactorings somewhere in the `docs/` in ALS repository?
* See other comments in files, but feel free to merge the PR if you want

File: source/ada/lsp-lal_utils.adb (hunk 1 above)
Comment (joaopsazevedo) [reply source/ada/lsp-lal_utils.adb:515]: Yes, I'll remove it.

File: source/protocol/lsp-messages.ads (hunk 2 above)
Comment (joaopsazevedo) [reply source/protocol/lsp-messages.ads:1257]: Agree. I'll add those.

File: source/protocol/lsp-messages.adb (hunk 3 above)
Comment (joaopsazevedo) [reply source/protocol/lsp-messages.adb:1135]: Agree, that would definitely allow us to auto generate the Read/Write subprograms for these types, so I think we should do that. I'll work on that and update this pull request.

**Pull request merged**